Feat/sled

bonedaddy opened the pull request.

bonedaddy committed:
add low-level zero-copy pack/unpack of hlcts

bonedaddy committed:
add low level unpack/pack of document data

bonedaddy committed:
add sled storage backend

bonedaddy committed:
enable all tokio features

bonedaddy committed:
add a trait to enable returning default keyspace

bonedaddy committed:
misc fixes

bonedaddy committed:
update the examples to use sled db

bonedaddy committed:
code formatting

bonedaddy committed:
fix lints

bonedaddy committed:
wip

bonedaddy committed:
wip

bonedaddy committed:
wip

bonedaddy committed:
wip

bonedaddy committed:
wip

bonedaddy committed:
add intensive python tests

bonedaddy committed:
remove useless changes

bonedaddy committed:
misc cleanup

bonedaddy committed:
lint

bonedaddy committed:
test debugging

bonedaddy committed:
fix integration issues

bonedaddy committed:
revert change

bonedaddy committed:
add 2 node cluster test for sled

bonedaddy committed:
fmt

bonedaddy committed:
apply lints

bonedaddy committed:
fmt

bonedaddy committed:
add additional tests which leverage sled

bonedaddy committed:
fix debug log

ChillFish8 commented on datacake-cluster/src/storage.rs:
```diff
@@ -752,6 +784,13 @@ pub mod test_suite {
             .mark_as_tombstone(KEYSPACE, doc_3.id, doc_3.last_updated)
             .await
             .expect("Delete documents from store.");
+        // NOTE: i manually added this call as the test seems to expect the entry in the keyspace
+        //       to have been removed
+        storage
+            .remove_tombstones(KEYSPACE, [3_u64].into_iter())
+            .await
+            .unwrap();
+
```
This is incorrect, this should work with the `mark_as_tombstone` without the need for this, Marking a document as a tombstone means the data can be removed entirely but the metadata must stay available when fetching with the metadata methods in order to rebuild the state.

It's incorrect to do this as the store will *not* call remove_tomestones unless it has observed state changes from the node issuing the removal. This is a metadata method as part of the CRDT purge system.

ChillFish8 commented on datacake-cluster/src/storage.rs:
```diff
@@ -736,6 +761,13 @@ pub mod test_suite {
             )
             .await
             .expect("Merk documents as tombstones");
+        // NOTE: i manually added this call as the test seems to expect the entry in the keyspace
+        //       to have been removed
+        storage
+            .remove_tombstones(KEYSPACE, [1_u64, 2_u64, 4_u64].into_iter())
+            .await
+            .unwrap();
+
```
Mentioned it below, but you can't do this, tombstones can potentially live forever theoretically, the data can be removed but the metadata must stay, and any `get_x` requests afterwards should act as if the data is removed.

ChillFish8 commented on datacake-cluster/src/storage.rs:
```diff
@@ -711,6 +727,14 @@ pub mod test_suite {
             .mark_as_tombstone(KEYSPACE, doc_2.id, doc_2.last_updated)
             .await
             .expect("Mark document as tombstone.");
+
+        // NOTE: i manually added this call as the test seems to expect the entry in the keyspace
+        //       to have been removed
+        storage
+            .remove_tombstones(KEYSPACE, [2_u64].into_iter())
+            .await
+            .unwrap();
+
```
Same as below

bonedaddy committed:
remove changes as per pr comments

bonedaddy committed:
correct how tombstones are handled

bonedaddy committed:
remove useless test suite changes

bonedaddy commented:
@ChillFish8 thank you for taking the time to review these changes! I have addresses the issues highlighted by your comments please take another look when you have a moment.

I do have a question:
> Marking a document as a tombstone means the data can be removed entirely but the metadata must stay available when fetching with the metadata methods in order to rebuild the state.

So if I understand correctly, when a document is marked as tombstoned, any `get_x` calls which would return the document should filter out the tombstoned document, but any calls that retrieve the metadata shouldn't filter out the metadata for any tombstoned documents?

ChillFish8 commented:
That is correct, any of the `get_*` methods should just return None or act like it doesn't exist if it's been marked as a tombstone.

bonedaddy commented:
@ChillFish8 tyvm, would this PR be accepted upstream? if so is there anything else i should do first?

ChillFish8 commented:
In it's current state probably not, there's a lot I'd wanted changed to merge it upstream, I'll add the reviews tomorrow as I don't quite have time today :)

ChillFish8 commented:
To add to the comments:

- Please do not modify the existing examples to use the sled system, feel free to add another example, maybe something a bit different using sled, but there's no reason to remove the sqlite demo.
- All sled DB operations should be run in a background thread to prevent blocking the runtimes, since this is a library we don't want to implicitly block runtimes.
